contract/src/agoric-basics.contract.js

```diff
@@ -22,76 +22,118 @@
 import { Far } from '@endo/far';
 import { M, getCopyBagEntries } from '@endo/patterns';
 import { AssetKind } from '@agoric/ertp/src/amountMath.js';
-import { AmountShape } from '@agoric/ertp/src/typeGuards.js';
 import { atomicRearrange } from '@agoric/zoe/src/contractSupport/atomicTransfer.js';
 import '@agoric/zoe/exported.js';
+import { AmountMath, AmountShape } from '@agoric/ertp';
 
 const { Fail, quote: q } = assert;
 
 // #region bag utilities
-/** @type { (xs: bigint[]) => bigint } */
-const sum = xs => xs.reduce((acc, x) => acc + x, 0n);
+/**
+ *
+ * @param {import('@endo/patterns').CopyBag} bag
+ * @param {[key: string]: {tradePrice: Amount, maxTickets: bigint}} inventory
+ * @returns {boolean}
+ */
+export const hasInventory = (bag, inventory) => {
+  const entries = getCopyBagEntries(bag);
+  for (const [k, ct] of entries) {
+    if (k in inventory) {
+      const { maxTickets } = inventory[k];
+      if (ct > maxTickets) {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+/**
+ *
+ * @param {Amount} amount
+ * @param {number} n
+ * @returns {Amount}
+ */
+const multiply = (amount, n) => {
+  const arr = Array.from({ length: n });
+  return arr.reduce(
+    (sum, _) => AmountMath.add(amount, sum),
+    AmountMath.make(amount.brand, 0n),
+  );
+};
+
+/**
+ *
+ * @param {Amount} sum
+ * @param {[string, bigint]} entry
+ * @param {[key: string]: {tradePrice: Amount, maxTickets: bigint}} inventory
+ * @returns {Amount}
+ */
+const addMultiples = (sum, entry, inventory) => {
+  const multiple = multiply(inventory[entry[0]].tradePrice, Number(entry[1]));
+  return AmountMath.add(multiple, sum);
+};
 
 /**
+ *
  * @param {import('@endo/patterns').CopyBag} bag
- * @returns {bigint[]}
+ * @param {[key: string]: {tradePrice: Amount, maxTickets: bigint}} inventory
+ * @returns {Amount}
  */
-const bagCounts = bag => {
+export const bagPrice = (bag, inventory) => {
   const entries = getCopyBagEntries(bag);
-  return entries.map(([_k, ct]) => ct);
+  const brand = Object.values(inventory)[0].tradePrice.brand;
+  return entries.reduce(
+    (sum, entry) => addMultiples(sum, entry, inventory),
+    // TODO: a better way to create empty amount
+    AmountMath.make(brand, 0n),
+  );
 };
 // #endregion
 
 /**
  * In addition to the standard `issuers` and `brands` terms,
  * this contract is parameterized by terms for price and,
- * optionally, a maximum number of items sold for that price (default: 3).
+ * optionally, a maximum number of tickets sold for that price (default: 3).
  *
  * @typedef {{
- *   tradePrice: Amount;
- *   maxItems?: bigint;
+ *   inventory: {[key: string]: {tradePrice: Amount, maxTickets: bigint}}
  * }} AgoricBasicsTerms
  */
 
-export const meta = {
-  customTermsShape: M.splitRecord(
-    { tradePrice: AmountShape },
-    { maxItems: M.bigint() },
-  ),
-};
-// compatibility with an earlier contract metadata API
-export const customTermsShape = meta.customTermsShape;
```

Comment:
`customTermsShape` remains relevant. It might be ok to leave it out for now, but latent bugs are likely from not checking the shape of `inventory` at contract startup.

Reply:
I found it useful to chase down a problem in something I was trying, so I might as well share...

```diff
modified   contract/src/agoric-basics-proposal.js
@@ -67,7 +67,16 @@ export const startAgoricBasicsContract = async permittedPowers => {
   const istIssuer = await istIssuerP;
   const istBrand = await istBrandP;
 
-  const terms = { tradePrice: AmountMath.make(istBrand, 25n * CENT) };
+  const i1 = AmountMath.make(istBrand, 1n * IST_UNIT);
+  const i2 = AmountMath.make(istBrand, 2n * IST_UNIT);
+  const i3 = AmountMath.make(istBrand, 3n * IST_UNIT);
+  const terms = {
+    inventory: {
+      frontRow: { tradePrice: i3, maxTickets: 3n },
+      middleRow: { tradePrice: i2, maxTickets: 3n },
+      lastRow: { tradePrice: i1, maxTickets: 3n },
+    },
+  };
```